File: modules/agent/src/serializers/job.py
"""Serializers for job-related models."""

from typing import Dict, Any, Optional

from lib.date_utils import format_date, format_datetime, format_display_date
from serializers.common import contact_to_dict


def extract_company_from_orm(job) -> tuple[str, str]:
    """Extract company name and type directly from ORM."""
    if not job.lead or not job.lead.account:
        return "", "Organization"

    if job.lead.account.organizations:
        return job.lead.account.organizations[0].name, "Organization"

    if job.lead.account.individuals:
        ind = job.lead.account.individuals[0]
        first_name = ind.first_name or ""
        last_name = ind.last_name or ""
        company = f"{last_name}, {first_name}".strip(", ")
        return company, "Individual"

    return "", "Organization"


def get_account_contacts(job) -> list:
    """Get contacts from job's account (org or individual)."""
    if not job.lead or not job.lead.account:
        return []

    if job.lead.account.organizations:
        return job.lead.account.organizations[0].contacts or []

    if job.lead.account.individuals:
        return job.lead.account.individuals[0].contacts or []

    return []


def get_salary_info(job, job_dict: dict) -> tuple[Optional[str], Optional[int]]:
    """Get salary range and ID from job."""
    if job.salary_range_ref:
        return job.salary_range_ref.label, job.salary_range_ref.id
    if job_dict:
        return job_dict.get("salary_range"), None
    return job.salary_range, None


def get_industries(job) -> list:
    """Get industry names from job's account."""
    if not job.lead or not job.lead.account or not job.lead.account.industries:
        return []
    return [ind.name for ind in job.lead.account.industries]


def get_status_name(job) -> str:
    """Get status name from job lead."""
    if not job.lead or not job.lead.current_status:
        return "Applied"
    return job.lead.current_status.name


def get_project_ids(job) -> list:
    """Get project IDs from job lead."""
    if not job.lead or not job.lead.projects:
        return []
    return [p.id for p in job.lead.projects]


def job_to_list_response(job) -> Dict[str, Any]:
    """Convert job ORM to response dictionary - optimized for list/table view."""
    company, _ = extract_company_from_orm(job)
    status = get_status_name(job)
    created_at = job.lead.created_at if job.lead else None

    return {
        "id": str(job.id),
        "account": company,
        "job_title": job.job_title or "",
        "status": status,
        "description": job.description,
        "resume": format_datetime(job.resume_date),
        "formatted_resume_date": format_display_date(job.resume_date),
        "job_url": job.job_url,
        "created_at": format_datetime(created_at),
        "formatted_created_at": format_display_date(created_at),
        "updated_at": format_datetime(job.lead.updated_at if job.lead else None),
        "formatted_updated_at": format_display_date(job.lead.updated_at if job.lead else None),
    }


def job_to_detail_response(job) -> Dict[str, Any]:
    """Convert job ORM to full response dictionary - for detail/edit views."""
    company, company_type = extract_company_from_orm(job)
    status = get_status_name(job)
    created_at = job.lead.created_at if job.lead else None
    salary_range, salary_range_id = get_salary_info(job, {})
    project_ids = get_project_ids(job)
    contacts = [contact_to_dict(c) for c in get_account_contacts(job)]
    industry = get_industries(job)

    return {
        "id": str(job.id),
        "account": company,
        "account_type": company_type,
        "job_title": job.job_title or "",
        "date": format_date(created_at),
        "formatted_date": format_display_date(created_at),
        "status": status,
        "job_url": job.job_url,
        "salary_range": salary_range,
        "salary_range_id": salary_range_id,
        "description": job.description,
        "resume": format_datetime(job.resume_date),
        "formatted_resume_date": format_display_date(job.resume_date),
        "source": job.lead.source if job.lead else "manual",
        "created_at": format_datetime(created_at),
        "updated_at": format_datetime(job.lead.updated_at if job.lead else None),
        "contacts": contacts,
        "industry": industry,
        "project_ids": project_ids,
    }

File: modules/agent/internal/serializers/job.go
package serializers

import (
	"time"

	"github.com/pina-colada-co/agent-go/internal/models"
)

func formatDisplayDate(t time.Time) string {
	if t.IsZero() {
		return ""
	}
	return t.Format("01/02/2006")
}

func formatDisplayDatePtr(t *time.Time) string {
	if t == nil || t.IsZero() {
		return ""
	}
	return t.Format("01/02/2006")
}

func formatISODate(t time.Time) string {
	if t.IsZero() {
		return ""
	}
	return t.Format("2006-01-02")
}

func formatDateTimePtr(t *time.Time) string {
	if t == nil || t.IsZero() {
		return ""
	}
	return t.Format(time.RFC3339)
}

// JobListResponse represents a job in list view
type JobListResponse struct {
	ID                   int64      `json:"id"`
	JobTitle             string     `json:"job_title"`
	JobURL               *string    `json:"job_url"`
	Description          *string    `json:"description"`
	Resume               string     `json:"resume"`
	ResumeDate           *time.Time `json:"resume_date"`
	FormattedResumeDate  string     `json:"formatted_resume_date"`
	SalaryRange          *string    `json:"salary_range"`
	SalaryRangeID        *int64     `json:"salary_range_id"`
	Status               *string    `json:"status"`
	Account              string     `json:"account"`
	AccountType          string     `json:"account_type"`
	Date                 string     `json:"date"`
	FormattedDate        string     `json:"formatted_date"`
	CreatedAt            time.Time  `json:"created_at"`
	FormattedCreatedAt   string     `json:"formatted_created_at"`
	UpdatedAt            time.Time  `json:"updated_at"`
	FormattedUpdatedAt   string     `json:"formatted_updated_at"`
	Source               string     `json:"source"`
	ProjectIDs           []int64    `json:"project_ids"`
}

// JobDetailResponse represents a job in detail view
type JobDetailResponse struct {
	JobListResponse
	Lead       *LeadResponse   `json:"lead"`
	Projects   []ProjectBrief  `json:"projects"`
	LeadStatus StatusResponse  `json:"lead_status,omitempty"`
}

// LeadResponse represents a lead
type LeadResponse struct {
	ID            int64        `json:"id"`
	Source        *string      `json:"source"`
	Type          string       `json:"type"`
	CurrentStatus *StatusBrief `json:"current_status"`
	TenantID      *int64       `json:"tenant_id"`
	CreatedAt     time.Time    `json:"created_at"`
	UpdatedAt     time.Time    `json:"updated_at"`
}

// OrganizationBrief represents org summary
type OrganizationBrief struct {
	ID   int64  `json:"id"`
	Name string `json:"name"`
}

// IndividualBrief represents individual summary
type IndividualBrief struct {
	ID        int64   `json:"id"`
	FirstName string  `json:"first_name"`
	LastName  string  `json:"last_name"`
	Email     *string `json:"email"`
}

// StatusBrief represents status summary
type StatusBrief struct {
	ID       int64   `json:"id"`
	Name     string  `json:"name"`
	Category *string `json:"category"`
}

// StatusResponse represents a full status
type StatusResponse struct {
	ID          int64   `json:"id"`
	Name        string  `json:"name"`
	Description *string `json:"description"`
	Category    *string `json:"category"`
	IsTerminal  bool    `json:"is_terminal"`
}

// StatusToResponse converts a Status model to response
func StatusToResponse(status *models.Status) StatusResponse {
	if status == nil {
		return StatusResponse{}
	}
	return StatusResponse{
		ID:          status.ID,
		Name:        status.Name,
		Description: status.Description,
		Category:    status.Category,
		IsTerminal:  status.IsTerminal,
	}
}

// ProjectBrief represents project summary
type ProjectBrief struct {
	ID   int64  `json:"id"`
	Name string `json:"name"`
}

// JobToListResponse converts a Job model to list response
func JobToListResponse(job *models.Job) JobListResponse {
	resp := JobListResponse{
		ID:                  job.ID,
		JobTitle:            job.JobTitle,
		JobURL:              job.JobURL,
		Description:         job.Description,
		Resume:              formatDateTimePtr(job.ResumeDate),
		ResumeDate:          job.ResumeDate,
		FormattedResumeDate: formatDisplayDatePtr(job.ResumeDate),
		SalaryRange:         job.SalaryRange,
		SalaryRangeID:       job.SalaryRangeID,
		Account:             "",
		AccountType:         "Organization",
		Source:              "manual",
	}

	if job.Lead == nil {
		return resp
	}

	resp.Date = formatISODate(job.Lead.CreatedAt)
	resp.FormattedDate = formatDisplayDate(job.Lead.CreatedAt)
	resp.CreatedAt = job.Lead.CreatedAt
	resp.FormattedCreatedAt = formatDisplayDate(job.Lead.CreatedAt)
	resp.UpdatedAt = job.Lead.UpdatedAt
	resp.FormattedUpdatedAt = formatDisplayDate(job.Lead.UpdatedAt)

	if job.Lead.Source != nil {
		resp.Source = *job.Lead.Source
	}

	if job.Lead.CurrentStatus != nil {
		resp.Status = &job.Lead.CurrentStatus.Name
	}

	if job.Lead.Account != nil {
		resp.Account, resp.AccountType = extractCompanyFromAccount(job.Lead.Account)
	}

	if job.Lead.Projects != nil {
		resp.ProjectIDs = make([]int64, len(job.Lead.Projects))
		for i, p := range job.Lead.Projects {
			resp.ProjectIDs[i] = p.ID
		}
	}

	return resp
}

// extractCompanyFromAccount gets company name and type from account
func extractCompanyFromAccount(account *models.Account) (string, string) {
	if account == nil {
		return "", "Organization"
	}

	if len(account.Organizations) > 0 {
		return account.Organizations[0].Name, "Organization"
	}

	if len(account.Individuals) == 0 {
		return "", "Organization"
	}

	ind := account.Individuals[0]
	firstName := ind.FirstName
	lastName := ind.LastName

	if lastName != "" && firstName != "" {
		return lastName + ", " + firstName, "Individual"
	}
	if lastName != "" {
		return lastName, "Individual"
	}
	if firstName != "" {
		return firstName, "Individual"
	}

	return "", "Organization"
}

// JobToDetailResponse converts a Job model to detail response
func JobToDetailResponse(job *models.Job) JobDetailResponse {
	resp := JobDetailResponse{
		JobListResponse: JobToListResponse(job),
	}

	if job.Lead == nil {
		return resp
	}

	resp.Lead = leadToResponse(job.Lead)

	if job.Lead.Projects == nil {
		return resp
	}

	resp.Projects = make([]ProjectBrief, len(job.Lead.Projects))
	for i, p := range job.Lead.Projects {
		resp.Projects[i] = ProjectBrief{ID: p.ID, Name: p.Name}
	}

	return resp
}

func leadToResponse(lead *models.Lead) *LeadResponse {
	if lead == nil {
		return nil
	}

	resp := &LeadResponse{
		ID:        lead.ID,
		Source:    lead.Source,
		Type:      lead.Type,
		TenantID:  lead.TenantID,
		CreatedAt: lead.CreatedAt,
		UpdatedAt: lead.UpdatedAt,
	}

	if lead.CurrentStatus != nil {
		resp.CurrentStatus = &StatusBrief{
			ID:       lead.CurrentStatus.ID,
			Name:     lead.CurrentStatus.Name,
			Category: lead.CurrentStatus.Category,
		}
	}

	return resp
}
